### PR TITLE
DISCUSSION: nil values for properties

### PR DESCRIPTION
**Pull request by jkeiser**
This is just a place for discussion and comments. Do not merge.

### CODE REVIEW

File: nil-discussion.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+   - `property :x` has `nil` as its default value.
+
+2. The default value of a property must be settable (the copy rule).
+
+   Corollary: Setters and default values are validated, so nothing can get *into*
+   the property unless it is valid.
+
+   - `x("10")` for `property :x, Integer` is INVALID.
+   - `property :x, Integer, default: "10"` is INVALID.
+   - `property :x, Integer, default: lazy { "10" }` is INVALID.
```
Comment (coderanger) [inline nil-discussion.md]: I would accept a system where this isn't checked but I agree it should be invalid at least in the abstract.

File: nil-discussion.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+```
+property :x, String
+property :x, String, default: nil # ValidationError when declaring this
+```
+
+This is not intuitive. It's not terrible in my opinion, as the failure of
+intuition would be revealed immediately on declaration with a clear error
+message.
+
+### Properties with no default or `default: nil` are valid
```
Comment (coderanger) [inline nil-discussion.md]: I agree with your ranking, this and the above seem like the best, and I tend toward this one fur the lower surprise and improved compat with existing code.

**Comment by coderanger** [comment]
:+1: We should probably chuck this in an informational RFC if/when we settle on a solution that that future-us has an easy reference for why there is so much legacy plumbing. Or we could just YOLO and release Chef 13 next month :trollface:

**Comment by jkeiser** [comment]
Yeah, I honestly don't think anyone outside of you, Lamont and I really care about nil handling :) An informational RFC would be good for sure!

I lean slightly towards the other solution, myself, just for conceptual parsimony--it burns me that `default: lazy { thing_that_can_return_nil }` would validate `nil` but `default: nil` would not--but I'm sure not going to throw an exception either way.

**Comment by stevendanna** [comment]
>  Yeah, I honestly don't think anyone outside of you, Lamont and I really care about nil handling :) An informational RFC would be good for sure!

I don't think that is true (I feel like these issues have been the cause of many subtle bugs I've had to fix in cookbooks over time); however, I definitely think it is true that you three are likely a submajority of the people who know the fallout of these various paths with relation to Chef internals :)

**Comment by coderanger** [comment]
I feel like validating lazy defaults can have a little more conceptual wiggle room as it's already an middling-to-advanced feature and we don't have to worry as much about footguns compared to the simple cases.

File: nil-discussion.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+This passes all of the rules, and is simple. If the default is not specified,
+then we are giving you a default, and we should take our own steps to make sure
+it is a valid one.
+
+The only issue is that these two would NOT be equivalent:
+
+```
+property :x, String
+property :x, String, default: nil # ValidationError when declaring this
```
Comment (stevendanna) [inline nil-discussion.md]: This being a ValidationError is OK with me in theory but is exactly the kind of minor annoyance that leads to frustration I think.  "If the 'default default' is nil, why can't I set it to nil?", my imagined user says.

File: nil-discussion.md (hunk 2 above)
Comment (stevendanna) [inline nil-discussion.md]: :+1:

**Comment by jkeiser** [comment]
I have fully implemented this, using the scheme where `default: nil` validates `nil` (but that will be easy to change) here: https://github.com/chef/chef/pull/4331

Warning when default values are invalid actually found a couple of bugs in the current code that could have had real implications (they are bugs I added post-12.6 when I converted packages over, but bugs nonetheless).  Happy with that one!

**Comment by lamont-granquist** [comment]
closing since i think any discussion needs to happen in #4331

**Pull request closed without merging**